client/landing/stepper/declarative-flow/internals/hooks/use-step-navigation-with-tracking/index.ts

```diff
@@ -28,21 +37,70 @@ export const useStepNavigationWithTracking = ( {
 		useSelect( ( select ) => ( select( ONBOARD_STORE ) as OnboardSelect ).getIntent(), [] ) ?? '';
 	const tracksEventPropsFromFlow = flow.useTracksEventProps?.();
 
+	const handleRecordStepNavigation = useCallback(
+		( {
+			event,
+			providedDependencies,
+			additionalProps,
+		}: Omit< RecordStepNavigationParams, 'step' | 'intent' | 'flow' | 'variant' > ) => {
+			recordStepNavigation( {
+				event,
+				intent,
+				flow: flow.name,
+				step: currentStepRoute,
+				variant: flow.variantSlug,
+				providedDependencies,
+				additionalProps,
+			} );
+		},
+		[ intent, currentStepRoute, flow ]
+	);
+
 	return useMemo(
 		() => ( {
-			...stepNavigation,
 			submit: ( providedDependencies: ProvidedDependencies = {}, ...params: string[] ) => {
-				recordSubmitStep(
+				handleRecordStepNavigation( {
+					event: STEPPER_TRACKS_EVENT_STEP_NAV_SUBMIT,
 					providedDependencies,
-					intent,
-					flow.name,
-					currentStepRoute,
-					flow.variantSlug,
-					tracksEventPropsFromFlow?.[ STEPPER_TRACKS_EVENT_STEP_NAV_SUBMIT ]
-				);
+					additionalProps: tracksEventPropsFromFlow?.[ STEPPER_TRACKS_EVENT_STEP_NAV_SUBMIT ],
+				} );
 				stepNavigation.submit?.( providedDependencies, ...params );
 			},
+			exitFlow: ( to: string ) => {
+				handleRecordStepNavigation( {
+					event: STEPPER_TRACKS_EVENT_STEP_NAV_EXIT_FLOW,
+					additionalProps: {
+						to,
+						...( tracksEventPropsFromFlow?.[ STEPPER_TRACKS_EVENT_STEP_NAV_EXIT_FLOW ] ?? {} ),
+					},
+				} );
+				stepNavigation.exitFlow?.( to );
+			},
+			goBack: () => {
+				handleRecordStepNavigation( {
+					event: STEPPER_TRACKS_EVENT_STEP_NAV_GO_BACK,
+					additionalProps: tracksEventPropsFromFlow?.[ STEPPER_TRACKS_EVENT_STEP_NAV_GO_BACK ],
+				} );
+				stepNavigation.goBack?.();
+			},
+			goNext: () => {
+				handleRecordStepNavigation( {
+					event: STEPPER_TRACKS_EVENT_STEP_NAV_GO_NEXT,
+					additionalProps: tracksEventPropsFromFlow?.[ STEPPER_TRACKS_EVENT_STEP_NAV_GO_NEXT ],
+				} );
+				stepNavigation.goNext?.();
+			},
+			goToStep: ( step: string ) => {
+				handleRecordStepNavigation( {
+					event: STEPPER_TRACKS_EVENT_STEP_NAV_GO_TO,
+					additionalProps: {
+						to: step,
+						...( tracksEventPropsFromFlow?.[ STEPPER_TRACKS_EVENT_STEP_NAV_GO_TO ] ?? {} ),
+					},
+				} );
+				stepNavigation.goToStep?.( step );
+			},
```

Comment:
Indeed! and good catch. It could lead to unexpected side-effects, like a step conditioning on them being present.

It sounds safe to define them only if the flow does.

Reply:
It should be addressed now. Need to test this more thoroughly. 

Also relying on `useStepNavigation` being a required prop on the flow, which is what the type is (so should be safe, unless some non TS flow pops up without it - if that will even work)